statsmodels/stats/dist_dependence_measures.py

```diff
@@ -0,0 +1,496 @@
+        random vector, i.e., each column represents a different component of
+        the random vector `x`.
+    y : array_like, 1-D or 2-D
+        Same as `x`, but only the number of observation has to match that of
+        `x`. If `y` is 2-D note that the number of columns of `y` (i.e., the
+        number of components in the random vector) does not need to match
+        the number of columns in `x`.
+    B : int
+        The number of iterations when evaluating the null distribution.
+    n : Number of observations found in each of `x` and `y`.
+    stats: namedtuple
+        The result obtained from calling ``calc_statistics(x, y)``.
+
+    Returns
+    -------
+    test_statistic : float
+        The empirical test statistic.
+    pval: float
+        The empirical p-value.
+
+    """
+    B = int(B) if B else int(np.floor(200 + 5000 / n))
+    empirical_dist = _get_test_statistic_distribution(x, y, B)
+    pval = 1 - np.searchsorted(
+        sorted(empirical_dist), stats.test_statistic
+    ) / len(empirical_dist)
+    test_statistic = stats.test_statistic
+
+    return test_statistic, pval
+
+
+def _asymptotic_pvalue(stats):
+    r"""Calculate the p-value based on an approximation of the distribution of
+    the test statistic under the null.
+
+    Parameters
+    ----------
+    stats: namedtuple
+        The result obtained from calling ``calc_statistics(x, y)``.
+
+    Returns
+    -------
+    test_statistic : float
+        The test statistic.
+    pval: float
+        The asymptotic p-value.
+
+    """
+    test_statistic = np.sqrt(stats.test_statistic / stats.S)
+    pval = (1 - norm.cdf(test_statistic)) * 2
+
+    return test_statistic, pval
+
+
+def _get_test_statistic_distribution(x, y, B):
+    r"""
+    Parameters
+    ----------
+    x : array_like, 1-D or 2-D
+        If `x` is 1-D than it is assumed to be a vector of observations of a
+        single random variable. If `x` is 2-D than the rows should be
+        observations and the columns are treated as the components of a
+        random vector, i.e., each column represents a different component of
+        the random vector `x`.
+    y : array_like, 1-D or 2-D
+        Same as `x`, but only the number of observation has to match that of
+        `x`. If `y` is 2-D note that the number of columns of `y` (i.e., the
+        number of components in the random vector) does not need to match
+        the number of columns in `x`.
+    B : int
+        The number of iterations to perform when evaluating the null
+        distribution.
+
+    Returns
+    -------
+    emp_dist : array_like
+        The empirical distribution of the test statistic.
+
+    """
+    y = y.copy()
+    emp_dist = np.zeros(B)
+    x_dist = squareform(pdist(x, "euclidean"))
+
+    for i in range(B):
+        np.random.shuffle(y)
+        emp_dist[i] = calc_statistics(x, y, x_dist=x_dist).test_statistic
+
+    return emp_dist
+
+
+def calc_statistics(x, y, x_dist=None, y_dist=None):
```

Comment:
Good looking PR.  The only quibble I have is about naming.  The function names should be as clear as possible without referencing the module name.  Perhaps `distance_statistics` here (open to suggestions).

Reply:
changed